### ISSUE TITLE
Failed to deploy on Heroku.

### ISSUE THREAD

**Issue by hetul-patel** [openvinotoolkit/model_server#323]
Hi, I am trying to deploy ovms on heroku but getting this error. I also tried assigning --rest_port=$PORT but
cannot resolve the issue. I get this error **_Web process failed to bind to $PORT within 60 seconds of launch_**

**Dockerfile**

```
FROM openvino/model_server:latest as build_image

# Set where models should be stored in the container
ENV MODEL_BASE_PATH /opt/ml
ENV MODEL_NAME facedetect

COPY models/facedetect /opt/ml/facedetect

# Run this command to start ovms
COPY --chown=admin:admin ovms_entrypoint.sh /usr/bin/ovms_entrypoint.sh
ENTRYPOINT []
CMD ["/usr/bin/ovms_entrypoint.sh"]
```

**ovms_entrypoint.sh**

```
#!/bin/bash

echo $PORT
/ovms/bin/ovms --model_name=${MODEL_NAME} --model_path=${MODEL_BASE_PATH}/${MODEL_NAME} --log_level DEBUG "$@"
```

**Log file from Heroku**
```
2020-10-27T20:32:02.458949+00:00 heroku[web.1]: Starting process with command `/usr/bin/ovms_entrypoint.sh`
2020-10-27T20:32:05.229729+00:00 app[web.1]: 20344
2020-10-27T20:32:05.363836+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] CLI parameters passed to ovms server
2020-10-27T20:32:05.363844+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] model_path: /opt/ml/facedetect
2020-10-27T20:32:05.363865+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] model_name: facedetect
2020-10-27T20:32:05.363881+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] batch_size: 0
2020-10-27T20:32:05.363881+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] shape: 
2020-10-27T20:32:05.363881+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] model_version_policy: 
2020-10-27T20:32:05.363882+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] nireq: 0
2020-10-27T20:32:05.363885+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] target_device: CPU
2020-10-27T20:32:05.363885+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] plugin_config: 
2020-10-27T20:32:05.363888+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] gRPC port: 9178
2020-10-27T20:32:05.363932+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] REST port: 0
2020-10-27T20:32:05.363932+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] REST workers: 24
2020-10-27T20:32:05.363933+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] gRPC workers: 1
2020-10-27T20:32:05.363933+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] gRPC channel arguments: 
2020-10-27T20:32:05.363937+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] log level: DEBUG
2020-10-27T20:32:05.363937+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] log path: 
2020-10-27T20:32:05.364030+00:00 app[web.1]: [2020-10-27 20:32:05.363] [serving] [debug] Batch size set: false, shape set: false
2020-10-27T20:32:05.364185+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [debug] Currently registered versions count:0
2020-10-27T20:32:05.364189+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [info] Getting model from /opt/ml/facedetect
2020-10-27T20:32:05.364189+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [info] Model downloaded to /opt/ml/facedetect
2020-10-27T20:32:05.364214+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [info] Will add model: facedetect; version: 1 ...
2020-10-27T20:32:05.364311+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [info] Loading model: facedetect, version: 1, from path: /opt/ml/facedetect/1, with target device: CPU ...
2020-10-27T20:32:05.364321+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [info] STATUS CHANGE: Version 1 of model facedetect status change. New status: ( "state": "START", "error_code": "OK" )
2020-10-27T20:32:05.364325+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [debug] setLoading: facedetect - 1 (previous state: START) -> error: OK
2020-10-27T20:32:05.364326+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [info] STATUS CHANGE: Version 1 of model facedetect status change. New status: ( "state": "LOADING", "error_code": "OK" )
2020-10-27T20:32:05.364329+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [debug] Getting model files from path:/opt/ml/facedetect/1
2020-10-27T20:32:05.364808+00:00 app[web.1]: [2020-10-27 20:32:05.364] [serving] [debug] Try reading model file:/opt/ml/facedetect/1/face-detection-0200.xml
2020-10-27T20:32:05.595846+00:00 app[web.1]: [2020-10-27 20:32:05.594] [serving] [debug] Network shape - (1,3,256,256); Final shape - (1,3,256,256)
2020-10-27T20:32:05.597750+00:00 app[web.1]: [2020-10-27 20:32:05.594] [serving] [info] Input name: image; mapping_name: ; shape: 1 3 256 256 ; precision: FP32, layout:NCHW
2020-10-27T20:32:05.597752+00:00 app[web.1]: [2020-10-27 20:32:05.594] [serving] [debug] model: facedetect, version: 1; reshaping inputs is not required
2020-10-27T20:32:05.597753+00:00 app[web.1]: [2020-10-27 20:32:05.594] [serving] [info] Output name: detection_out ; mapping name: ; shape: 1 1 200 7  ; precision: FP32, layout:NCHW
2020-10-27T20:33:02.683257+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
2020-10-27T20:33:02.738447+00:00 heroku[web.1]: Stopping process with SIGKILL
2020-10-27T20:33:03.255409+00:00 heroku[web.1]: Process exited with status 137
2020-10-27T20:33:03.303212+00:00 heroku[web.1]: State changed from starting to crashed
```

**Comment by dtrawins** [comment]
@hetul-patel  OVMS starts by default the gRPC service on port 9178. You can change it be adding `--port` parameter. `--rest_port` is optional to enable also REST API endpoint.
Based on the logs there is a problem in binding the port for gRPC endpoint so 9178.
Could you check if something is blocking that port or change it with --port option?

**Comment by hetul-patel** [comment]
@dtrawins I tried assigning port using --port argument but same error occurs.

**ovms_entrypoint.sh**

```
#!/bin/bash

echo heroku port is $PORT
/ovms/bin/ovms --port $PORT --model_name=${MODEL_NAME} --model_path=${MODEL_BASE_PATH}/${MODEL_NAME} --log_level DEBUG "$@"
```

**Log file from Heroku**
```
2020-10-28T16:40:00.495214+00:00 heroku[web.1]: Starting process with command `/usr/bin/ovms_entrypoint.sh`
2020-10-28T16:40:03.715084+00:00 app[web.1]: heroku port is 47414
2020-10-28T16:40:04.151008+00:00 app[web.1]: [2020-10-28 16:40:04.147] [serving] [debug] CLI parameters passed to ovms server
2020-10-28T16:40:04.151050+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] model_path: /opt/ml/facedetect
2020-10-28T16:40:04.151050+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] model_name: facedetect
2020-10-28T16:40:04.151051+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] batch_size: 0
2020-10-28T16:40:04.151051+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] shape: 
2020-10-28T16:40:04.151052+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] model_version_policy: 
2020-10-28T16:40:04.151052+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] nireq: 0
2020-10-28T16:40:04.151052+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] target_device: CPU
2020-10-28T16:40:04.151052+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] plugin_config: 
2020-10-28T16:40:04.151053+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] gRPC port: 47414
2020-10-28T16:40:04.151053+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] REST port: 0
2020-10-28T16:40:04.151054+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] REST workers: 24
2020-10-28T16:40:04.151054+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] gRPC workers: 1
2020-10-28T16:40:04.151054+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] gRPC channel arguments: 
2020-10-28T16:40:04.151055+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] log level: DEBUG
2020-10-28T16:40:04.151055+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] log path: 
2020-10-28T16:40:04.151055+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] Batch size set: false, shape set: false
2020-10-28T16:40:04.151056+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] Currently registered versions count:0
2020-10-28T16:40:04.151056+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [info] Getting model from /opt/ml/facedetect
2020-10-28T16:40:04.151056+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [info] Model downloaded to /opt/ml/facedetect
2020-10-28T16:40:04.151057+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [info] Will add model: facedetect; version: 1 ...
2020-10-28T16:40:04.151058+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [info] Loading model: facedetect, version: 1, from path: /opt/ml/facedetect/1, with target device: CPU ...
2020-10-28T16:40:04.151063+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [info] STATUS CHANGE: Version 1 of model facedetect status change. New status: ( "state": "START", "error_code": "OK" )
2020-10-28T16:40:04.151064+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] setLoading: facedetect - 1 (previous state: START) -> error: OK
2020-10-28T16:40:04.151064+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [info] STATUS CHANGE: Version 1 of model facedetect status change. New status: ( "state": "LOADING", "error_code": "OK" )
2020-10-28T16:40:04.151064+00:00 app[web.1]: [2020-10-28 16:40:04.149] [serving] [debug] Getting model files from path:/opt/ml/facedetect/1
2020-10-28T16:40:04.151065+00:00 app[web.1]: [2020-10-28 16:40:04.150] [serving] [debug] Try reading model file:/opt/ml/facedetect/1/face-detection-0200.xml
2020-10-28T16:40:04.793823+00:00 app[web.1]: [2020-10-28 16:40:04.793] [serving] [debug] Network shape - (1,3,256,256); Final shape - (1,3,256,256)
2020-10-28T16:40:04.793842+00:00 app[web.1]: [2020-10-28 16:40:04.793] [serving] [info] Input name: image; mapping_name: ; shape: 1 3 256 256 ; precision: FP32, layout:NCHW
2020-10-28T16:40:04.793842+00:00 app[web.1]: [2020-10-28 16:40:04.793] [serving] [debug] model: facedetect, version: 1; reshaping inputs is not required
2020-10-28T16:40:04.793843+00:00 app[web.1]: [2020-10-28 16:40:04.793] [serving] [info] Output name: detection_out ; mapping name: ; shape: 1 1 200 7  ; precision: FP32, layout:NCHW
2020-10-28T16:41:01.202381+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
2020-10-28T16:41:01.230057+00:00 heroku[web.1]: Stopping process with SIGKILL
2020-10-28T16:41:01.507500+00:00 heroku[web.1]: Process exited with status 137
2020-10-28T16:41:01.555239+00:00 heroku[web.1]: State changed from starting to crashed
```

**Comment by dtrawins** [comment]
@hetul-patel  It looks like a problem related to Heroku.
Perhaps this will help in finding a solution https://stackoverflow.com/questions/15693192/heroku-node-js-error-web-process-failed-to-bind-to-port-within-60-seconds-of 
OVMS binds the port to all interfaces.

**Comment by rasapala** [comment]
@hetul-patel Please let us know if you need more help.

**Comment by hetul-patel** [comment]
@rasapala I tried the solution provided by @dtrawins but the same error occurs.

**Comment by Mohamad1994HD** [comment]
From the log

"Web process failed to bind to $PORT within 60 seconds of launch"

Maybe I am wrong but why is it to $PORT instead of a value since you want to pass the value of it, not the string itself.

Is it passing '$PORT' or passing the value of PORT?

**Comment by rasapala** [comment]
@hetul-patel this looks like the app setup problem on heroku side.
@Mohamad1994HD  suggested using PORT not $PORT.
In the link provided by @dtrawins  you should fix app setup by using 0.0.0.0 instead of localhost or adding the || port number to your app setup. From the logs you have sent it looks like ovms is working as expected. Not sure we can help on Heroku setup here.
